### PR TITLE
Add NULL values

### LINKED ISSUE #7: Adapt format file
Align `.clang-format` with Hyrise counterpart.

### LINKED ISSUE #8: Handle `NULL` values in segments
- [x] add `NullValue` as data type for `AllTypeVariant`, define `AllTypeVariant{}` as `NULL_VALUE` and add `variant_is_null(...)`
- [x] add `std::optional<T> get_typed_value(const ChunkOffset chunk_offset) const` to `ValueSegment`/`DictionarySegment`
- [x] add `const std::vector<bool>& null_values() const` to `ValueSegment`
- [x] add tests for null handling in dictionaries, scans
- [x] (minor) replace `throw std::logic_error(...)` with `Fail(...)`

### PR DESCRIPTION
**Pull request by dey4ss**
This PR adds NULL values to the segment interfaces, as well as tests for their handling in segments and the scan operator. Furthermore, the format script is updated to the Hyrise version. Additional changes: removes superfluous includes, fails all unimplemented methods, moves skeletons to cpp files. The `EXPLICITLY_DECLARE_DATA_TYPES(DictionarySegment)` macro call silences a linker warning for (somewhat recent) clang versions.

closes #7 
closes #8

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -18,7 +18,7 @@ endif()
 add_definitions("-DSOURCE_PATH_SIZE=${SOURCE_PATH_SIZE}")
 
 # Global flags and include directories
-add_compile_options(-std=c++1z -pthread -Wall -Wextra -pedantic -Werror -Wno-unused-parameter)
+add_compile_options(-std=c++20 -pthread -Wall -Wextra -pedantic -Werror -Wno-unused-parameter)
```
Comment (Bouncner) [inline src/CMakeLists.txt:21]: Uff, good catch.

File: scripts/format.sh [hunk 2]
```diff
@@ -2,7 +2,8 @@
 
 unamestr=$(uname)
 if [[ "$unamestr" == 'Darwin' ]]; then
-	format_cmd="clang-format -i -style=file '{}'"
+    clang_format="$(brew --prefix llvm)/bin/clang-format"
```
Comment (dey4ss) [inline scripts/format.sh:5]: update to ARM Mac

File: scripts/lint.sh [hunk 3]
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-find src \( -iname "*.cpp" -o -iname "*.hpp" \) -print0 | parallel --null --no-notice -j 100% --nice 17 python3 ./third_party/cpplint/cpplint.py --verbose=0 --extensions=hpp,cpp --counting=detailed --filter=-legal/copyright,-whitespace/newline,-runtime/references,-build/c++11,-build/include_subdir --linelength=120 {} 2\>\&1 \| grep -v \'\^Done processing\' \| grep -v \'\^Total errors found: 0\' \; test \${PIPESTATUS[0]} -eq 0
+find src \( -iname "*.cpp" -o -iname "*.hpp" \) -print0 | parallel --null --no-notice -j 100% --nice 17 python3 ./third_party/cpplint/cpplint.py --verbose=0 --extensions=hpp,cpp --counting=detailed --filter=-legal/copyright,-whitespace/newline,-runtime/references,-build/c++11,-build/include_subdir,-build/include_what_you_use --linelength=120 {} 2\>\&1 \| grep -v \'\^Done processing\' \| grep -v \'\^Total errors found: 0\' \; test \${PIPESTATUS[0]} -eq 0
```
Comment (dey4ss) [inline scripts/lint.sh:3]: This rule requested the bloat of includes that where transitively included anyways. We also exclude this rule in Hyrise.

File: src/lib/all_type_variant.hpp [hunk 4]
```diff
@@ -81,6 +95,13 @@ using AllTypeVariant = detail::AllTypeVariant;
  * @{
  */
 
+#define EXPLICIT_DECLARATION(r, template_class, type) extern template class template_class<type>;
+
+// Explicitly declares the given template class for all types in DATA_TYPES (used in .hpp)
+#define EXPLICITLY_DECLARE_DATA_TYPES(template_class)                           \
+  BOOST_PP_SEQ_FOR_EACH(EXPLICIT_DECLARATION, template_class, data_types_macro) \
+  static_assert(true, "End call of macro with a semicolon")
```
Comment (dey4ss) [inline src/lib/all_type_variant.hpp:103]: Required for not having the linker warning for DictionarySegment.

File: src/lib/null_value.hpp [hunk 5]
```diff
@@ -0,0 +1,65 @@
+#pragma once
```
Comment (dey4ss) [inline src/lib/null_value.hpp:1]: Basically 1:1 copy from Hyrise.

File: src/test/storage/dictionary_segment_test.cpp [hunk 6]
```diff
@@ -31,18 +30,21 @@ TEST_F(StorageDictionarySegmentTest, CompressSegmentString) {
 
   auto dict_segment = std::dynamic_pointer_cast<DictionarySegment<std::string>>(segment);
```
Comment (dey4ss) [inline src/test/storage/dictionary_segment_test.cpp:31]: I don't really get the purpose of not just doing
```c++
auto dict_segment = std::make_shared<DictionarySegment<std::string>>(value_segment_str);
```

File: src/test/storage/chunk_test.cpp [hunk 7]
```diff
@@ -43,8 +39,8 @@ TEST_F(StorageChunkTest, AddValuesToChunk) {
   EXPECT_EQ(chunk.size(), 4u);
 
   if constexpr (HYRISE_DEBUG) {
```
Comment (dey4ss) [inline src/test/storage/chunk_test.cpp:41]: Pick here: `HYRISE_DEBUG`, and the binaries are called `hyrise...`. Change to opossum for the full opossum experience?

File: src/lib/type_cast.cpp [hunk 8]
```diff
@@ -1,9 +1,9 @@
 #include "type_cast.hpp"
 
-#include <string>
-
 namespace opossum {
 
-std::string to_string(const AllTypeVariant& x) { return boost::lexical_cast<std::string>(x); }
+std::string to_string(const AllTypeVariant& x) {
```
Comment (Bouncner) [inline src/lib/type_cast.cpp:5]: x -> variant?

File: src/lib/utils/string_utils.cpp [hunk 9]
```diff
@@ -31,7 +31,8 @@ std::vector<std::string> split_string_by_delimiter(const std::string& str, char
 
 std::string trim_source_file_path(const std::string& path) {
   const auto src_pos = path.find("/src/");
-  if (src_pos == std::string::npos) return path;
+  if (src_pos == std::string::npos)
+    return path;
```
Comment (Bouncner) [inline src/lib/utils/string_utils.cpp:35]: Brackets.

File: src/test/base_test.cpp [hunk 10]
```diff
@@ -40,7 +33,8 @@ BaseTest::Matrix BaseTest::_table_to_matrix(const Table& table) {
     const auto chunk = table.get_chunk(chunk_id);
 
     // an empty table's chunk might be missing actual segments
-    if (chunk->size() == 0) continue;
+    if (chunk->size() == 0)
```
Comment (Bouncner) [inline src/test/base_test.cpp:36]: Brackets

File: src/test/base_test.cpp [hunk 11]
```diff
@@ -40,7 +33,8 @@ BaseTest::Matrix BaseTest::_table_to_matrix(const Table& table) {
     const auto chunk = table.get_chunk(chunk_id);
 
     // an empty table's chunk might be missing actual segments
```
Comment (Bouncner) [inline src/test/base_test.cpp:35]: // An ... segmentsDOT

File: src/test/base_test.cpp (hunk 10 above)
Comment (Bouncner) [reply src/test/base_test.cpp:36]: Do we still miss the format rule to enforce brackets?

Comment (Bouncner) [reply src/test/base_test.cpp:36]: (curly braces)

File: src/test/lib/all_type_variant_test.cpp [hunk 12]
```diff
@@ -30,10 +25,18 @@ TYPED_TEST(AllTypeVariantTest, GetExtractsExactNumericalValue) {
   }
 
   for (auto value_in : values) {
```
Comment (Bouncner) [inline src/test/lib/all_type_variant_test.cpp:27]: const

File: src/test/storage/chunk_test.cpp (hunk 7 above)
Comment (Bouncner) [reply src/test/storage/chunk_test.cpp:41]: If it's easy to change, I'd say yes.

File: src/test/storage/value_segment_test.cpp [hunk 13]
```diff
@@ -33,18 +28,26 @@ TEST_F(StorageValueSegmentTest, AddValueOfSameType) {
   EXPECT_EQ(double_value_segment.size(), 1u);
 }
 
-TEST_F(StorageValueSegmentTest, AddValueOfDifferentType) {
+TEST_F(StorageValueSegmentTest, ApendValueOfDifferentType) {
```
Comment (Bouncner) [inline src/test/storage/value_segment_test.cpp:31]: Apend

File: src/test/storage/dictionary_segment_test.cpp (hunk 6 above)
Comment (Bouncner) [reply src/test/storage/dictionary_segment_test.cpp:31]: I have no idea.

File: src/test/base_test.cpp (hunk 10 above)
Comment (dey4ss) [reply src/test/base_test.cpp:36]: https://github.com/hyrise/hyrise/blob/f7ad570cc30208c71593c0d7f609e34c0f60decb/.clang-format#L59

I think it's currently only enforced by clang-tidy.

File: src/lib/type_cast.cpp (hunk 8 above)
Comment (dey4ss) [reply src/lib/type_cast.cpp:5]: The method is not even declared in the header and removing the .cpp does not break compilation. I would just remove the whole file.

File: src/test/lib/all_type_variant_test.cpp (hunk 12 above)
Comment (dey4ss) [reply src/test/lib/all_type_variant_test.cpp:27]: even `const auto&` :D

File: src/test/base_test.cpp (hunk 10 above)
Comment (Bouncner) [reply src/test/base_test.cpp]: Ah true.

File: src/test/storage/value_segment_test.cpp [hunk 14]
```diff
@@ -54,4 +57,28 @@ TEST_F(StorageValueSegmentTest, MemoryUsage) {
...
+  EXPECT_FALSE(int_value_segment.is_null(0));
+  EXPECT_TRUE(int_value_segment.is_null(1));
+
+  EXPECT_EQ(int_value_segment.get(0), 1);
+  EXPECT_THROW(int_value_segment.get(1), std::logic_error);
+
+  EXPECT_EQ(int_value_segment.get_typed_value(0), 1);
+  EXPECT_EQ(int_value_segment.get_typed_value(1), std::nullopt);
+
+  EXPECT_EQ(int_value_segment.values().size(), int_value_segment.null_values().size());
```
Comment (mweisgut) [inline src/test/storage/value_segment_test.cpp:76]: This line asserts that `values` and `null_values` have the same number of elements. In addition, I would check that the number of elements in `values` and `null_values` is 2 in this example. With line 76, checking the size of only one of them should be sufficient.

**Review by Bouncner — APPROVED** [review]
Minor things.

**Review by mweisgut — APPROVED** [review]

**Review by Bouncner — APPROVED** [review]
